### PR TITLE
Get metadata_snapshot from the Replicated* table instead of Distribut…

### PR DESCRIPTION
**Pull request by sunset3000**
…ed table when execute query on distributed table.

I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

**Changelog category (leave one):**
- Improvement

**Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):**

Fix the issue that when excute select query on distributed table, the query plan has unessary sorting step

**Detailed description / Documentation draft:**
When execute query on distributed table, in the following case. the query plan has extra ordering steps. 

Table DDL:
```
CREATE table event3 (
`_indextime` DateTime64(3) DEFAULT now() CODEC(DoubleDelta, LZ4),
_time ALIAS occur_time,
...

) Engine = ReplicatedMergeTree('/clickhouse/tables/event3/{shard}', '{replica}')
-- ) Engine = MergeTree
ORDER BY (toStartOfMinute(occur_time),event_name, src_address)
PARTITION BY toYYYYMMDD(occur_time)
```
The sorting key is ORDER BY (toStartOfMinute(occur_time),event_name, src_address)

Corresponding Distributed Table

```
CREATE TABLE default.event3_all ON CLUSTER `cluster_2shards_2replicas`
AS default.event3
ENGINE = Distributed(cluster_2shards_2replicas, default, event3, rand());
```
Non distributed Query:

`explain select event_name from event3 order by toStartOfMinute(occur_time) desc;`

Query plain:

┌─explain───────────────────────────────────────────────────────────────────────┐
│ Expression (Projection) │
│ FinishSorting │
│ Expression (Before ORDER BY) │
│ SettingQuotaAndLimits (Set limits and quota after reading from storage) │
│ Union │
│ ReverseRows │
│ ReadFromStorage (MergeTree with order) │
│ ReverseRows │
│ ReadFromStorage (MergeTree with order) │
│ ReverseRows │
│ ReadFromStorage (MergeTree with order) │
│ ReverseRows │
│ ReadFromStorage (MergeTree with order) │
│ ReverseRows │
│ ReadFromStorage (MergeTree with order) │
│ ReverseRows │
│ ReadFromStorage (MergeTree with order) │
│ ReverseRows │
│ ReadFromStorage (MergeTree with order) │
│ ReverseRows │
│ ReadFromStorage (MergeTree with order) │
│ ReverseRows │
│ ReadFromStorage (MergeTree with order) │
│ ReverseRows │
│ ReadFromStorage (MergeTree with order) │
│ ReverseRows │
│ ReadFromStorage (MergeTree with order) │
│ ReverseRows │
│ ReadFromStorage (MergeTree with order) │
│ ReverseRows │
│ ReadFromStorage (MergeTree with order) │
│ ReverseRows │
│ ReadFromStorage (MergeTree with order) │
│ ReverseRows │
│ ReadFromStorage (MergeTree with order) │
│ ReverseRows │
│ ReadFromStorage (MergeTree with order) │
└───────────────────────────────────────────────────────────────────────────────┘
 

The query plan has 'RevserRows' steps only, which excute quite fast without merge sorting. 

Distributed Query

`explain select event_name from event3_all order by toStartOfMinute(occur_time) desc;`
Query Plan

┌─explain───────────────────────────────────────────────────────────────────────────────────┐
│ Expression (Projection)                                                                   │
│   MergingSorted (Merge sorted streams for ORDER BY)                                       │
│     SettingQuotaAndLimits (Set limits and quota after reading from storage)               │
│       Union                                                                               │
│         Expression (Convert block structure for query from local replica)                 │
│           MergingSorted (Merge sorted streams for ORDER BY)                               │
│             MergeSorting (Merge sorted blocks for ORDER BY)                               │
│               PartialSorting (Sort each block for ORDER BY)                               │
│                 Expression (Before ORDER BY)                                              │
│                   SettingQuotaAndLimits (Set limits and quota after reading from storage) │
│                     ReadFromStorage (MergeTree)                                           │
│         ReadFromPreparedSource (Read from remote replica)                                 │
└───────────────────────────────────────────────────────────────────────────────────────────┘
 The query plan has PartialSorting, MergeSorting, MergingSorted steps, which slow down the query quite a lot.

Analysis:

The metadata of Distributed Engine does not have sorting_key information, therefore the ReadInOrderOptimizer cannot optimized the query properly. The fix should add the missing sorting_key info when create Distributed storage.

Expected Result:

for distributed query, it should be the same as non-distributed query without extra soring steps.  

By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.

Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 1]
```diff
@@ -323,7 +324,22 @@ InterpreterSelectQuery::InterpreterSelectQuery(
...
-            metadata_snapshot = storage->getInMemoryMetadataPtr();
+        {
+            if (storage->getName() != "Distributed")
+            {
+                metadata_snapshot = storage->getInMemoryMetadataPtr();
+            }
+            else
+            {
+                const StorageDistributed * storage_distributed = static_cast<const StorageDistributed *>(storage.get());
+                StoragePtr storage_replicated = DatabaseCatalog::instance().getTable(
```
Comment (alexey-milovidov) [inline src/Interpreters/InterpreterSelectQuery.cpp:335]: It only works if table is remote table is present in local shard.

Comment (sunset3000) [reply src/Interpreters/InterpreterSelectQuery.cpp:335]: I afraid yes. Do you have any suggestion how to make it work when remote table is not present in local shard? Thanks a lot.

**Comment by alexey-milovidov** [comment]
I don't have any idea how to implement it properly, let's ask @CurtizJ, @KochetovNicolai.

File: src/Interpreters/InterpreterSelectQuery.cpp (hunk 1 above)
Comment (sunset3000) [reply src/Interpreters/InterpreterSelectQuery.cpp]: I have fixed the failed stateless tests by switching to previous behavior in some cases.

**Comment by alexey-milovidov** [comment]
@CurtizJ The change is too ad-hoc. It should be implemented in some other way.

**Pull request closed without merging**